File: services/llm-gateway/src/provider_deepseek.py
from __future__ import annotations

import json
import os
import ssl
import urllib.error
import urllib.request
from typing import Any, Callable

try:
    import certifi
    _CAFILE: str | None = certifi.where()
except ImportError:
    _CAFILE = None

Transport = Callable[[urllib.request.Request, float], bytes]


def chat_completion(
    messages: list[dict[str, str]],
    *,
    model: str,
    base_url: str | None = None,
    api_key: str | None = None,
    timeout: float = 30.0,
    transport: Transport | None = None,
) -> str:
    api_key = api_key or os.environ.get("DEEPSEEK_API_KEY")
    if not api_key:
        raise RuntimeError("DEEPSEEK_API_KEY is not set")

    base_url = (base_url or os.environ.get("DEEPSEEK_BASE_URL") or "https://api.deepseek.com").rstrip("/")
    payload = {
        "model": model,
        "messages": messages,
        "stream": False,
    }
    body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    request = urllib.request.Request(
        f"{base_url}/chat/completions",
        data=body,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        method="POST",
    )

    try:
        raw = (transport or _default_transport)(request, timeout)
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"DeepSeek API error {exc.code}: {detail}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"DeepSeek API network error: {exc.reason}") from exc

    data: dict[str, Any] = json.loads(raw.decode("utf-8"))
    return data["choices"][0]["message"]["content"]


def _default_transport(request: urllib.request.Request, timeout: float) -> bytes:
    ctx = None
    if _CAFILE:
        ctx = ssl.create_default_context(cafile=_CAFILE)
    with urllib.request.urlopen(request, timeout=timeout, context=ctx) as response:
        return response.read()

File: services/llm-gateway/ts-src/providerDeepseek.ts
export type TransportResponse = {
  status: number;
  body: string;
};

export type Transport = (
  url: string,
  options: { method: string; headers: Record<string, string>; body: string },
  timeout: number,
) => Promise<TransportResponse>;

export class HttpError extends Error {
  statusCode: number;
  responseBody: string;
  constructor(statusCode: number, responseBody: string) {
    super(`HTTP ${statusCode}`);
    this.statusCode = statusCode;
    this.responseBody = responseBody;
    this.name = "HttpError";
  }
}

export type ChatCompletionOptions = {
  model: string;
  base_url?: string;
  api_key?: string;
  timeout?: number;
  transport?: Transport;
};

function pyDictGet<T>(d: Record<string, unknown>, key: string): T {
  if (!(key in d)) {
    const err = new Error(`'${key}'`);
    err.name = "KeyError";
    throw err;
  }
  return d[key] as T;
}

function pyListGet<T>(arr: T[], index: number): T {
  if (index >= arr.length) {
    const err = new Error("list index out of range");
    err.name = "IndexError";
    throw err;
  }
  return arr[index] as T;
}

function parseDeepseekJson(text: string): unknown {
  try {
    return JSON.parse(text);
  } catch (e: unknown) {
    const err = new Error((e as Error).message);
    err.name = "JSONDecodeError";
    throw err;
  }
}

function pyRuntimeError(message: string): Error {
  const err = new Error(message);
  err.name = "RuntimeError";
  return err;
}

async function defaultTransport(
  url: string,
  options: { method: string; headers: Record<string, string>; body: string },
  timeout: number,
): Promise<TransportResponse> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeout * 1000);
  try {
    const resp = await fetch(url, { ...options, signal: controller.signal });
    const body = await resp.text();
    if (!resp.ok) {
      throw new HttpError(resp.status, body);
    }
    return { status: resp.status, body };
  } finally {
    clearTimeout(timer);
  }
}

import { deepseekApiKey, deepseekBaseUrl } from "./config.js";

export async function chatCompletion(
  messages: Array<{ role: string; content: string }>,
  options: ChatCompletionOptions,
): Promise<string> {
  const apiKey = options.api_key ?? deepseekApiKey();
  if (!apiKey) {
    throw pyRuntimeError("DEEPSEEK_API_KEY is not set");
  }

  const baseUrl = (options.base_url ?? deepseekBaseUrl()).replace(/\/+$/, "");
  const url = `${baseUrl}/chat/completions`;

  const payload: Record<string, unknown> = {
    model: options.model,
    messages: messages,
    stream: false,
  };
  const body = JSON.stringify(payload);

  const transportFn: Transport = options.transport ?? defaultTransport;

  let raw: string;
  try {
    const response = await transportFn(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`,
      },
      body,
    }, options.timeout ?? 30.0);
    raw = response.body;
  } catch (e: unknown) {
    if (e instanceof HttpError) {
      throw pyRuntimeError(`DeepSeek API error ${e.statusCode}: ${e.responseBody}`);
    }
    const msg = e instanceof Error ? e.message : String(e);
    throw pyRuntimeError(`DeepSeek API network error: ${msg}`);
  }

  const data = parseDeepseekJson(raw) as Record<string, unknown>;
  const choices = pyDictGet<Array<Record<string, unknown>>>(data, "choices");
  const first = pyListGet(choices, 0);
  const message = pyDictGet<Record<string, unknown>>(first, "message");
  const content = pyDictGet<string>(message, "content");
  return content;
}
